### PR TITLE
Implement missing share settings

### PR DESCRIPTION
**Pull request by allexzander**
### BEFORE
share_settings_before
### AFTER
share_settings_after

### CODE REVIEW

File: src/gui/filedetails/NCRadioButton.qml [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        anchors.leftMargin: 5
+        radius: root.radius
+        border.color: root.color
+        border.width: 1
+        Rectangle {
+            anchors.fill: parent
+            visible: root.checked
+            color: root.color
+            radius: root.radius
+            anchors.margins: 4
```
Comment (claucambra) [inline src/gui/filedetails/NCRadioButton.qml:38]: We should avoid magic numbers/colours, these should go in Style

File: src/gui/filedetails/NCRadioButton.qml [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+import QtQuick 2.15
+import QtQuick.Controls 2.15
+
+RadioButton {
+    id: root
+    property int indicatorItemWidth: 16
+    property int indicatorItemHeight: 16
+    property string color: "#000000"
+    readonly property int radius: 9
+    indicator: Rectangle {
```
Comment (claucambra) [inline src/gui/filedetails/NCRadioButton.qml:24]:
```suggestion
    readonly property int indicatorItemWidth: 16
    readonly property int indicatorItemHeight: 16
    readonly property string color: "#000000"
    readonly property int radius: 9
    
    indicator: Rectangle {
```

File: src/gui/filedetails/ShareDetailsPage.qml [hunk 3]
```diff
@@ -313,34 +311,128 @@ Page {
...
+                    indicator.height: moreMenu.indicatorItemWidth
+
+                    checkable: true
+                    checked: root.editingAllowed
+                    text: qsTr("Allow upload and editing")
+                    enabled: !root.isSharePermissionChangeInProgress
+
+                    onClicked: {
+                        root.toggleAllowEditing(checked);
+                    }
```
Comment (claucambra) [inline src/gui/filedetails/ShareDetailsPage.qml:331]:
```suggestion
                    onClicked: root.toggleAllowEditing(checked)
```

File: src/gui/filedetails/ShareDetailsPage.qml [hunk 4]
```diff
@@ -313,34 +311,128 @@ Page {
...
+                        id: permissionModeRadioButtonsGroup
+                    }
 
-                checkable: true
-                checked: root.editingAllowed
-                text: qsTr("Allow editing")
-                enabled: !root.waitingForEditingAllowedChange
-                visible: !root.isSecureFileDropLink
+                    NCRadioButton {
+                        property int permissionMode: ShareModel.ModeViewOnly
```
Comment (claucambra) [inline src/gui/filedetails/ShareDetailsPage.qml:356]:
```suggestion
                        readonly property int permissionMode: ShareModel.ModeViewOnly
```

File: src/gui/filedetails/ShareDetailsPage.qml [hunk 5]
```diff
@@ -313,34 +311,128 @@ Page {
...
+                        spacing: moreMenu.indicatorSpacing
+                        padding: moreMenu.itemPadding
+                        onClicked: root.permissionModeChanged(permissionMode);
+                    }
 
-                onClicked: {
-                    root.toggleAllowEditing(checked);
-                    root.waitingForEditingAllowedChange = true;
+                    NCRadioButton {
+                        property int permissionMode: ShareModel.ModeUploadAndEditing
```
Comment (claucambra) [inline src/gui/filedetails/ShareDetailsPage.qml:370]:
```suggestion
                        readonly property int permissionMode: ShareModel.ModeUploadAndEditing
```

File: src/gui/filedetails/ShareDetailsPage.qml [hunk 6]
```diff
@@ -313,34 +311,128 @@ Page {
...
+                        NCBusyIndicator {
+                            anchors.fill: parent
+                            visible: root.isSharePermissionChangeInProgress
+                            running: visible
+                            z: 1
+                        }
+                    }
+
+                    NCRadioButton {
+                        property int permissionMode: ShareModel.ModeFileDropOnly
```
Comment (claucambra) [inline src/gui/filedetails/ShareDetailsPage.qml:391]:
```suggestion
                        readonly property int permissionMode: ShareModel.ModeFileDropOnly
```

File: src/gui/filedetails/ShareDetailsPage.qml [hunk 7]
```diff
@@ -313,34 +311,128 @@ Page {
...
+                        spacing: moreMenu.indicatorSpacing
+                        padding: moreMenu.itemPadding
+                        indicator.width: moreMenu.indicatorItemWidth
+                        indicator.height: moreMenu.indicatorItemWidth
+                        checked: root.hideDownload
+                        text: qsTr("Hide download")
+                        enabled: !root.isHideDownloadInProgress
+                        onClicked: {
+                            root.toggleHideDownload(checked);
+                        }
```
Comment (claucambra) [inline src/gui/filedetails/ShareDetailsPage.qml:427]:
```suggestion
                        onClicked:  root.toggleHideDownload(checked)
```

File: src/gui/filedetails/sharemodel.cpp [hunk 8]
```diff
@@ -722,6 +785,18 @@ void ShareModel::slotSharePasswordSet(const QString &shareId)
     Q_EMIT dataChanged(shareModelIndex, shareModelIndex, { PasswordProtectEnabledRole, PasswordRole });
 }
 
+void ShareModel::slotHideDowloadSet(const QString &shareId)
```
Comment (claucambra) [inline src/gui/filedetails/sharemodel.cpp:788]:
```suggestion
void ShareModel::slotHideDownloadSet(const QString &shareId)
```

File: src/gui/filedetails/sharemodel.cpp [hunk 9]
```diff
@@ -867,6 +961,42 @@ void ShareModel::toggleShareNoteToRecipientFromQml(const QVariant &share, const
     toggleShareNoteToRecipient(ptr, enable);
 }
 
+void ShareModel::changePermissionMode(const QVariant &share, const SharePermissionsMode permissionMode)
```
Comment (claucambra) [inline src/gui/filedetails/sharemodel.cpp:964]: I believe the naming here should be consistent with the others

```suggestion
void ShareModel::changePermissionModeFromQml(const QVariant &share, const SharePermissionsMode permissionMode)
```

File: src/gui/sharemanager.cpp [hunk 10]
```diff
@@ -203,14 +203,16 @@ LinkShare::LinkShare(AccountPtr account,
     const QUrl &url,
     const QDate &expireDate,
     const QString &note,
-    const QString &label)
+    const QString &label,
+    bool hideDownload)
```
Comment (claucambra) [inline src/gui/sharemanager.cpp:207]:
```suggestion
    const bool hideDownload)
```

File: src/gui/sharemanager.cpp [hunk 11]
```diff
@@ -308,6 +320,15 @@ void LinkShare::slotLabelSet(const QJsonDocument &, const QVariant &label)
     }
 }
 
+void LinkShare::slotHideDownloadSet(const QJsonDocument &jsonDoc, const QVariant hideDownload)
```
Comment (claucambra) [inline src/gui/sharemanager.cpp:323]:
```suggestion
void LinkShare::slotHideDownloadSet(const QJsonDocument &jsonDoc, const QVariant &hideDownload)
```

File: src/gui/filedetails/sharemodel.h [hunk 12]
```diff
@@ -164,6 +187,8 @@ private slots:
     void handleSecureFileDropLinkShare();
     void handleLinkShare();
     void setupInternalLinkShare();
+    void setSharePermissionChangeInProgress(const QString &shareId, bool isInProgress);
+    void setHideDownloadEnabledChangeInProgress(const QString &shareId, bool isInProgress);
```
Comment (claucambra) [inline src/gui/filedetails/sharemodel.h:191]:
```suggestion
    void setSharePermissionChangeInProgress(const QString &shareId, const bool isInProgress);
    void setHideDownloadEnabledChangeInProgress(const QString &shareId, const bool isInProgress);
```

File: src/gui/sharemanager.cpp [hunk 13]
```diff
@@ -583,7 +604,8 @@ QSharedPointer<LinkShare> ShareManager::parseLinkShare(const QJsonObject &data)
         url,
         expireDate,
         note,
-        data.value("label").toString()));
+        data.value("label").toString(),
+        static_cast<bool>(data.value("hide_download").toInt())));
```
Comment (claucambra) [inline src/gui/sharemanager.cpp:608]: Why not just use `toBool`?

File: src/gui/sharemanager.h [hunk 14]
```diff
@@ -213,7 +215,8 @@ class LinkShare : public Share
         const QUrl &url,
         const QDate &expireDate,
         const QString &note,
-        const QString &label);
+        const QString &label,
+        bool hideDownload);
```
Comment (claucambra) [inline src/gui/sharemanager.h:219]:
```suggestion
        const bool hideDownload);
```

File: src/gui/sharemanager.h [hunk 15]
```diff
@@ -291,18 +299,25 @@ public slots:
...
     void nameSet();
     void labelSet();
+    void hideDownloadSet();
 
 private slots:
     void slotNoteSet(const QJsonDocument &, const QVariant &value);
     void slotExpireDateSet(const QJsonDocument &reply, const QVariant &value);
     void slotNameSet(const QJsonDocument &, const QVariant &value);
     void slotLabelSet(const QJsonDocument &, const QVariant &value);
+    void slotHideDownloadSet(const QJsonDocument &jsonDoc, const QVariant hideDownload);
```
Comment (claucambra) [inline src/gui/sharemanager.h:320]:
```suggestion
    void slotHideDownloadSet(const QJsonDocument &jsonDoc, const QVariant &hideDownload);
```

**Review by claucambra — CHANGES_REQUESTED** [review]

File: src/gui/sharemanager.cpp (hunk 13 above)
Comment (claucambra) [reply src/gui/sharemanager.cpp:609]: Fair enough, though maybe a better way of writing this would be

```suggestion
        data.value("hide_download").toInt() == 1);
```

File: src/gui/filedetails/NCRadioButton.qml [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+
+    indicator: Rectangle {
+        implicitWidth: root.indicatorItemWidth
+        implicitHeight: root.indicatorItemHeight
+        anchors.verticalCenter: parent.verticalCenter
+        anchors.left: parent.left
+        anchors.leftMargin: Style.radioButtonCustomMarginLeftOuter
+        radius: root.radius
+        border.color: root.color
+        border.width: 1
```
Comment (claucambra) [inline src/gui/filedetails/NCRadioButton.qml:34]: We still have a magic number here

**Review by claucambra — COMMENTED** [review]
A couple of small comments left from my previous review, though they are nitpicks

File: src/gui/filedetails/sharemodel.h [hunk 17]
```diff
@@ -79,6 +83,23 @@ class ShareModel : public QAbstractListModel
         ShareTypeSecureFileDropPlaceholderLink = Share::TypeSecureFileDropPlaceholderLink,
     };
     Q_ENUM(ShareType);
+    
+    enum SharedItemType {
```
Comment (mgallien) [inline src/gui/filedetails/sharemodel.h:87]: why not `enum class` ?

File: src/gui/filedetails/sharemodel.h [hunk 18]
```diff
@@ -79,6 +83,23 @@ class ShareModel : public QAbstractListModel
...
+        SharedItemTypeUndefined = -1,
+        SharedItemTypeFile,
+        SharedItemTypeFolder,
+        SharedItemTypeEncryptedFile,
+        SharedItemTypeEncryptedFolder,
+        SharedItemTypeEncryptedTopLevelFolder,
+    };
+    Q_ENUM(SharedItemType);
+    
+    enum SharePermissionsMode {
```
Comment (mgallien) [inline src/gui/filedetails/sharemodel.h:97]: same question ?

**Review by mgallien — COMMENTED** [review]
mostly questions
sometime the code is a bit hard to follow between what comes from the network jobs and what comes from user actions

File: src/gui/filedetails/sharemodel.cpp (hunk 8 above)
Comment (mgallien) [reply src/gui/filedetails/sharemodel.cpp:788]: @alex see

File: src/gui/filedetails/sharemodel.cpp (hunk 9 above)
Comment (mgallien) [reply src/gui/filedetails/sharemodel.cpp:964]: @alex ping

File: src/gui/sharemanager.cpp [hunk 19]
```diff
@@ -308,6 +320,16 @@ void LinkShare::slotLabelSet(const QJsonDocument &, const QVariant &label)
     }
 }
 
+void LinkShare::slotHideDownloadSet(const QJsonDocument &jsonDoc, const QVariant &hideDownload)
+{
+    Q_UNUSED(jsonDoc);
+    if (!hideDownload.isValid()) {
+        return;
+    }
+    _hideDownload = hideDownload.toBool();
```
[the remark this answers is not in the document]
Comment (allexzander) [reply src/gui/sharemanager.cpp:329]: @mgallien I can not, the UI gets broken if I do what you ask

File: src/gui/filedetails/sharemodel.cpp [hunk 20]
```diff
@@ -151,8 +157,21 @@ QVariant ShareModel::data(const QModelIndex &index, const int role) const
         return expireDateEnforcedForShare(share);
     case EnforcedMaximumExpireDateRole:
         return enforcedMaxExpireDateForShare(share);
-    case IsSecureFileDropLinkRole:
-        return _isSecureFileDropSupportedFolder && share->getPermissions().testFlag(OCC::SharePermission::SharePermissionCreate);
+    case CurrentPermissionModeRole: {
+        if (share->getPermissions() == OCC::SharePermission::SharePermissionCreate) {
+            return static_cast<int>(SharePermissionsMode::ModeFileDropOnly);
```
Comment (mgallien) [inline src/gui/filedetails/sharemodel.cpp:162]: I had missed that in my first review
should be
```
return QVariand::fromValue(SharePermissionsMode::ModeFileDropOnly);
```

**Review by mgallien — APPROVED** [review]
sounds good
did not do any local tests during review
